<a href="https://colab.research.google.com/github/Gayathri2526/AIML-learning-Projects/blob/main/Car-Price-Prediction/Car_Price_Prediction_Linear_Ridge_Lasso.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Car Price Prediction - Linear, Ridge and Lasso Regression

## 1. Load and Understand the Dataset

In [1]:
import pandas as pd

df = pd.read_csv("Car_Price_Prediction.csv")

df.head()

,Make,Model,Year,Engine Size,Mileage,Fuel Type,Transmission,Price
0,Honda,Model B,2015,3.9,74176,Petrol,Manual,30246.207931
1,Ford,Model C,2014,1.7,94799,Electric,Automatic,22785.747684
2,BMW,Model B,2006,4.1,98385,Electric,Manual,25760.290347
3,Honda,Model B,2015,2.6,88919,Electric,Automatic,25638.003491
4,Honda,Model C,2004,3.4,138482,Petrol,Automatic,21021.386657


In [2]:
# Check dataset size
print("Dataset shape:", df.shape)

# Check column information
df.info()

Dataset shape: (1000, 8)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Make          1000 non-null   object 
 1   Model         1000 non-null   object 
 2   Year          1000 non-null   int64  
 3   Engine Size   1000 non-null   float64
 4   Mileage       1000 non-null   int64  
 5   Fuel Type     1000 non-null   object 
 6   Transmission  1000 non-null   object 
 7   Price         1000 non-null   float64
dtypes: float64(2), int64(2), object(4)
memory usage: 62.6+ KB


In [3]:
# Check missing values
df.isnull().sum()

,0
Make,0
Model,0
Year,0
Engine Size,0
Mileage,0
Fuel Type,0
Transmission,0
Price,0


In [4]:
# Summary statistics
df.describe()

,Year,Engine Size,Mileage,Price
count,1000.000000,1000.000000,1000.00000,1000.000000
mean,2010.688000,2.798300,97192.48700,25136.615530
std,6.288577,1.024137,59447.31576,5181.401368
min,2000.000000,1.000000,56.00000,6704.953524
25%,2005.000000,1.900000,44768.75000,21587.878370
50%,2011.000000,2.800000,94411.50000,25189.325247
75%,2016.000000,3.700000,148977.75000,28806.368974
max,2021.000000,4.500000,199867.00000,41780.504635


## 2. Prepare Features and Target

In [5]:
# Separate input features and target variable

X = df.drop("Price", axis=1)
y = df["Price"]

print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (1000, 7)
Target shape: (1000,)


## 3. Encode Categorical Features

In [6]:
# Convert categorical columns into numerical columns using one-hot encoding

X = pd.get_dummies(X, drop_first=True)

print("Shape after encoding:", X.shape)
X.head()

Shape after encoding: (1000, 14)


,Year,Engine Size,Mileage,Make_BMW,Make_Ford,Make_Honda,Make_Toyota,Model_Model B,Model_Model C,Model_Model D,Model_Model E,Fuel Type_Electric,Fuel Type_Petrol,Transmission_Manual
0,2015,3.9,74176,False,False,True,False,True,False,False,False,False,True,True
1,2014,1.7,94799,False,True,False,False,False,True,False,False,True,False,False
2,2006,4.1,98385,True,False,False,False,True,False,False,False,True,False,True
3,2015,2.6,88919,False,False,True,False,True,False,False,False,True,False,False
4,2004,3.4,138482,False,False,True,False,False,True,False,False,False,True,False


## 4. Train/Test Split

In [7]:
from sklearn.model_selection import train_test_split

# Split the data: 80% training and 20% testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (800, 14)
X_test shape: (200, 14)
y_train shape: (800,)
y_test shape: (200,)


## 5. Feature Scaling

In [8]:
from sklearn.preprocessing import StandardScaler

# Create scaler
scaler = StandardScaler()

# Fit only on training data
X_train_scaled = scaler.fit_transform(X_train)

# Apply same scaling to test data
X_test_scaled = scaler.transform(X_test)

print("Training data shape:", X_train_scaled.shape)
print("Testing data shape:", X_test_scaled.shape)

Training data shape: (800, 14)
Testing data shape: (200, 14)


## 6. Linear Regression Model

In [9]:
from sklearn.linear_model import LinearRegression

# Create Linear Regression model
linear_model = LinearRegression()

# Train the model
linear_model.fit(X_train_scaled, y_train)

# Predict car prices using test data
y_pred_linear = linear_model.predict(X_test_scaled)

print("Linear Regression model trained successfully.")

Linear Regression model trained successfully.


## 7. Ridge Regression Model

In [10]:
from sklearn.linear_model import Ridge

# Create Ridge Regression model
ridge_model = Ridge(alpha=1.0)

# Train the model
ridge_model.fit(X_train_scaled, y_train)

# Predict car prices using test data
y_pred_ridge = ridge_model.predict(X_test_scaled)

print("Ridge Regression model trained successfully.")

Ridge Regression model trained successfully.


## 8. Lasso Regression Model

In [11]:
from sklearn.linear_model import Lasso

# Create Lasso Regression model
lasso_model = Lasso(alpha=1.0)

# Train the model
lasso_model.fit(X_train_scaled, y_train)

# Predict car prices using test data
y_pred_lasso = lasso_model.predict(X_test_scaled)

print("Lasso Regression model trained successfully.")

Lasso Regression model trained successfully.


## 9. Model Evaluation

In [12]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

# Function to calculate evaluation metrics
def evaluate_model(y_test, y_pred):
    mae = mean_absolute_error(y_test, y_pred)
    mse = mean_squared_error(y_test, y_pred)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, y_pred)

    return mae, mse, rmse, r2


# Linear Regression metrics
linear_mae, linear_mse, linear_rmse, linear_r2 = evaluate_model(
    y_test, y_pred_linear
)

# Ridge Regression metrics
ridge_mae, ridge_mse, ridge_rmse, ridge_r2 = evaluate_model(
    y_test, y_pred_ridge
)

# Lasso Regression metrics
lasso_mae, lasso_mse, lasso_rmse, lasso_r2 = evaluate_model(
    y_test, y_pred_lasso
)

print("Model evaluation completed successfully.")

Model evaluation completed successfully.


## 10. Model Comparison

In [13]:
# Create comparison table

results = pd.DataFrame({
    "Model": ["Linear Regression", "Ridge Regression", "Lasso Regression"],
    "MAE": [linear_mae, ridge_mae, lasso_mae],
    "MSE": [linear_mse, ridge_mse, lasso_mse],
    "RMSE": [linear_rmse, ridge_rmse, lasso_rmse],
    "R2": [linear_r2, ridge_r2, lasso_r2]
})

# Round values for easier reading
results = results.round(4)

results

,Model,MAE,MSE,RMSE,R2
0,Linear Regression,1810.5548,5.005471e+06,2237.2910,0.8171
1,Ridge Regression,1810.3238,5.003749e+06,2236.9062,0.8172
2,Lasso Regression,1809.8563,5.002216e+06,2236.5634,0.8172


## 11. Coefficient Comparison

In [14]:
# Compare coefficients of all three models

coefficients = pd.DataFrame({
    "Feature": X.columns,
    "Linear": linear_model.coef_,
    "Ridge": ridge_model.coef_,
    "Lasso": lasso_model.coef_
})

coefficients = coefficients.round(4)

coefficients

,Feature,Linear,Ridge,Lasso
0,Year,3223.7934,3219.5098,3222.5299
1,Engine Size,2033.4004,2030.8491,2032.2650
2,Mileage,-2901.2292,-2897.6099,-2900.3631
3,Make_BMW,36.1011,36.3026,33.9472
4,Make_Ford,51.5361,51.7316,49.3071
5,Make_Honda,-54.3840,-53.5791,-54.8447
6,Make_Toyota,12.8867,13.0127,10.6435
7,Model_Model B,-18.2209,-18.1292,-15.7710
8,Model_Model C,-31.9052,-31.6367,-29.4510
9,Model_Model D,124.3076,124.2201,125.2686


## 12. Conclusion

Lasso Regression performed slightly better than Linear and Ridge Regression because it achieved the lowest MAE, MSE, and RMSE while maintaining the highest R² score of approximately 0.8172. Ridge and Lasso regularization slightly changed the model coefficients, helping control model complexity. Overall, Lasso provided the best performance on the test data, although the differences between all three models were very small.